### PR TITLE
wallet: Avoid multiple BerkeleyBatch in DelAddressBook

### PR DESCRIPTION
**Pull request by promag**

### CODE REVIEW

**Review by meshcollider — COMMENTED** [review]
Code review ACK b2ce2b97501f8642b79da024dd485545b67f5533

**Comment by achow101** [comment]
In `CWallet::FindNonChangeParentOutput`, I think we should add an `AssertLockHeld(cs_wallet)` too.

**Comment by promag** [comment]
> In `CWallet::FindNonChangeParentOutput`, I think we should add an `AssertLockHeld(cs_wallet)` too.

Done. Also rebased now that #19289 is merged.

**Comment by achow101** [comment]
ACK abac4367607d8d2b628e4db6a9663c960bacdacc

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2346,6 +2346,7 @@ std::map<CTxDestination, std::vector<COutput>> CWallet::ListCoins() const
 
 const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int output) const
 {
+    AssertLockHeld(cs_wallet);
```
Comment (jonatack) [inline src/wallet/wallet.cpp:2349]: @achow101 per https://github.com/bitcoin/bitcoin/pull/19738#issuecomment-675602253 I'm curious why this should be added, as there is already an `AssertLockHeld(cs_wallet);` at the top of its caller, `ListCoins()`. For a future call from elsewhere?

**Review by jonatack — COMMENTED** [review]
ACK abac4367607d8d2b628e4db6a9663c960bacdacc

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/wallet.cpp:2349]: This was based on #19289 which then raised #19773, and @achow101 suggestion should be done there, `ListCoins` eventually calls `FindNonChangeParentOutput`, and like `IsTrusted`, was missing the assertion.

Comment (promag) [reply src/wallet/wallet.cpp:2349]: @jonatack I could split to a different commit though to be more correct, just LMK.

Comment (jonatack) [reply src/wallet/wallet.cpp:2349]: No worries, I was only trying to understand the locking. This is fine.

**Review by meshcollider — COMMENTED** [review]
re-utACK abac4367607d8d2b628e4db6a9663c960bacdacc

**Pull request merged**